zerver/lib/actions.py

```diff
@@ -789,22 +790,28 @@ def do_send_messages(messages_maybe_none):
         Message.objects.bulk_create([message['message'] for message in messages])
         ums = []  # type: List[UserMessage]
         for message in messages:
+            # These properties on the Message are set via
+            # render_markdown by code in the bugdown inline patterns
+            wildcard = message['message'].mentions_wildcard
+            mentioned_ids = message['message'].mentions_user_ids
+            ids_with_alert_words = message['message'].user_ids_with_alert_words
+            is_me_message = message['message'].is_me_message
+
             # Service bots (outgoing webhook bots and embedded bots) don't store UserMessage rows;
             # they will be processed later.
+            # Also we make sure UserMessage rows are not created got
+            # long_term_idle unless the message contains special flag for which
+            # we would rather create a UserMessage row now rather then later.
             ums_to_create = []
             for user_profile in message['active_recipients']:
                 # is_service_bot is derived from is_bot and bot_type, and both of these fields
                 # should have been pre-fetched.
-                if not user_profile.is_service_bot:
+                if not user_profile.is_service_bot and (
+                    not user_profile.long_term_idle or
+                    user_profile.id in mentioned_ids or
+                    is_me_message):
                     ums_to_create.append(UserMessage(user_profile=user_profile, message=message['message']))
 
-            # These properties on the Message are set via
-            # render_markdown by code in the bugdown inline patterns
-            wildcard = message['message'].mentions_wildcard
-            mentioned_ids = message['message'].mentions_user_ids
-            ids_with_alert_words = message['message'].user_ids_with_alert_words
-            is_me_message = message['message'].is_me_message
-
             for um in ums_to_create:
                 if um.user_profile.id == message['message'].sender.id and \
                         message['message'].sent_by_human():
```

Comment:
I think a cleaned end-to-end model might be to actually create `UserMessage` rows that would have had any flags set, done as a filter on the `ums_to_create_list` a bit further down in this function.  That way, if we create anything new that has flags, it'll automatically do the right thing without needing to change the conditionals.